Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -83,6 +84,31 @@ MatrixWorkspace::~MatrixWorkspace() {
   }
 }
 
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::SpectrumNumbers &&spectrumNumbers,
+    Indexing::DetectorIDs &&detectorIDs) {
+  setIndexTranslator(Indexing::IndexTranslator(std::move(spectrumNumbers),
+                                               std::move(detectorIDs)));
+}
+
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::IndexTranslator &&translator) {
+  if (translator.size() != getNumberHistograms())
+    throw std::runtime_error("MatrixWorkspace::setIndexTranslator: Translator "
+                             "size does not match number of histograms in "
+                             "workspace");
+
+  auto &spectrumNumbers = translator.spectrumNumbers();
+  auto &detectorIDs = translator.detectorIDs();
+  for (size_t i = 0; i < getNumberHistograms(); ++i) {
+    auto &spectrum = getSpectrum(i);
+    spectrum.setSpectrumNo(spectrumNumbers[i]);
+    if (!detectorIDs.empty())
+      spectrum.setDetectorIDs(
+          std::set<detid_t>(detectorIDs[i].begin(), detectorIDs[i].end()));
+  }
+}
```

Comment:
Basically what I am doing so far is a first step of refactoring, with the goal to move all `setSpectrumNo` and `setDetectorID` calls into the body of `setIndexTranslator`. Once that is done we can remove to respective members (`ISpectrum::m_specNo`, ...) and instead store the translator.